ENH: stats: Add crosstab function.

WarrenWeckesser opened the pull request:
crosstab creates a contingency table from a set of input
sequences.  In effect, it counts the occurrences of the
tuples in zip(*args), and arranges the result in an array.

WarrenWeckesser commented:
Some notes comparing the API to similar functions in Pandas and R...

Simple example, for comparison to Pandas and R:

    In [13]: x = [1, 3, 4, 5, 4, 5, 1, 4, 3, 3, 4, 3, 3]

    In [14]: y = [1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0]

    In [15]: (xvals, yvals), count = crosstab(x, y)

    In [16]: xvals
    Out[16]: array([1, 3, 4, 5])

    In [17]: yvals
    Out[17]: array([0, 1])

    In [18]: count
    Out[18]: 
    array([[0, 2],
           [3, 2],
           [2, 2],
           [1, 1]])


Pandas has the function [`crosstab`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.crosstab.html):

    In [13]: import pandas as pd

    In [14]: x = [1, 3, 4, 5, 4, 5, 1, 4, 3, 3, 4, 3, 3]

    In [15]: y = [1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0]

    In [16]: pd.crosstab([x], [y])
    Out[16]: 
    col_0  0  1
    row_0      
    1      0  2
    3      3  2
    4      2  2
    5      1  1


Probably the simplest function in R that corresponds to this function is [`table`](https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/table):

    > x <- c(1, 3, 4, 5, 4, 5, 1, 4, 3, 3, 4, 3, 3)
    > y <- c(1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0)
    > table(x, y)
       y
    x   0 1
      1 0 2
      3 3 2
      4 2 2
      5 1 1


The parameter name `levels` matches the use in the R function [`factor`](https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/factor) that creates a factor.  The `table` function accepts factors:

    > fx = factor(x, levels=c(1, 2, 3, 4, 5))
    > fy = factor(y)
    > table(fx, fy)
       fy
    fx  0 1
      1 0 2
      2 0 0
      3 3 2
      4 2 2
      5 1 1

Here's the corresponding operation with the new `crosstab` function:

    In [23]: (xvals, yvals), count = crosstab(x, y, levels=[[1, 2, 3, 4, 5], None])

    In [24]: xvals
    Out[24]: [1, 2, 3, 4, 5]

    In [25]: yvals
    Out[25]: array([0, 1])

    In [26]: count
    Out[26]: 
    array([[0, 2],
           [0, 0],
           [3, 2],
           [2, 2],
           [1, 1]])


Pandas' `crosstab` includes group-by and aggregation capabilities.  We can eventually add that to our version of `crosstab` if there is demand.   That is, instead of simply counting the occurrences of the unique combination of the input factors, it would compute a statistic from the values given in another array of numbers.  This would be somewhat related to the `binned_statistic` function.

WarrenWeckesser commented:
The one test failure on Travis-CI is not related to this PR.

rlucas7 commented on scipy/stats/tests/test_crosstab.py:
```diff
@@ -0,0 +1,75 @@
+
+
+def test_crosstab_extra_levels():
+    # The pair of values (-1, 3) will be ignored, because we explicitly
```
Is it worth working this concept into the docstring somewhere? I can see confused user posting on stackoverflow about this after it is out in the wild.

WarrenWeckesser replied:
I added an example to the docstring where `levels` is used to ignore an input value.  Do you think that is enough?

rlucas7 replied:
> I added an example to the docstring where `levels` is used to ignore an input value. Do you think that is enough?

yeah should be enough, thanks

rlucas7 commented on scipy/stats/_crosstab.py:
```diff
@@ -0,0 +1,154 @@
+        does not occur in the corresponding sequence in `levels`, that value
+        is ignored and not counted in the returned array `count`.
+
+    Returns
```
could a sparse return be accommodated for large inputs?

matthew-brett replied:
I guess the sparse return would need an extra input argument - you wouldn't want to get something sparse for some data, and not for others ...   Maybe we could add this later with a `sparse=True` keyword argument?

WarrenWeckesser replied:
That would be a nice enhancement, but let's leave it for a future PR.

rlucas7 replied:
yes, both good points. I hadn't considered the sparse + non-sparse complexity. I agree w/Warren, let's leave for a future PR if someone requests.

rlucas7 commented on scipy/stats/tests/test_crosstab.py:
```diff
@@ -0,0 +1,75 @@
+from scipy.stats import crosstab
+
+
+def test_crosstab_basic():
```
can we put all the tests into a `class Test_crosstab` class for pytest? makes it easier to organize and share test input/output as the tests cases evolve. Even if we don't strictly ***need*** it now, it is nice to have and doesn't break anything setup here.

rlucas7 replied:
I had missed that this is within the file of tests specific to `crosstab` so no. need for the pytest class.

rlucas7 commented on scipy/stats/tests/test_crosstab.py:
```diff
@@ -0,0 +1,75 @@
+    assert_array_equal(count, expected_count)
+
+
+def test_crosstab_basic_1d():
```
General question, given the similarity with `pandas` is it worthwhile to have a test a common input/output pair of the `crosstab` from this PR against what is in `pandas`?

WarrenWeckesser replied:
I don't really see the need.  We wouldn't actually use pandas in the test, so it would just be another result in a test case.

rlucas7 replied:
yeah agreed, resolving.

rlucas7 commented on scipy/stats/_crosstab.py:
```diff
@@ -0,0 +1,154 @@
+import numpy as np
+
+
+def crosstab(*args, levels=None):
```
No firm opinion on my end but did you consider a `**kwargs` instead of individual arguments after the args tuple unpacking? Just thinking that `**kwargs` might make the signature more extensible in future.

rlucas7 replied:
spoke w/Ralf about this one, I was thinking the use of `**kwargs` would make the signature more flexible but the price of that flexibility wouldn't really give any benefit and we'd lose the a bunch so the way the signature is here is fine with me.

WarrenWeckesser commented:
@rlucas7, thanks for the review!  Just letting you know that it might be a few days before I get back to this.

rgommers commented:
I forgot if or where I said this before, but: it's a little weird to have `crosstab` in the main namespace, and then have a separate `stats.contingency` namespace as well.

rlucas7 commented:
> I forgot if or where I said this before, but: it's a little weird to have `crosstab` in the main namespace, and then have a separate `stats.contingency` namespace as well.

Ralf brings up a good point here. Warren perhaps if you move the `crosstab` function under contingency it makes more sense. As far as the interoperability of `crosstab` output with existing contingency methods (e.g. `margins` and `expected_freq`) but it looks like those should work fine:
```
>>> from scipy.stats import crosstab
>>> x = [1, 3, 4, 5, 4, 5, 1, 4, 3, 3, 4, 3, 3]
>>> y = [1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0]
>>> (xvals, yvals), count = crosstab(x, y)
>>> from scipy.stats.contingency import margins
>>> xvals
array([1, 3, 4, 5])
>>> yvals
array([0, 1])
>>> count
array([[0, 2],
       [3, 2],
       [2, 2],
       [1, 1]])

>>> margins(count)
[array([[2],
       [5],
       [4],
       [2]]), array([[6, 7]])]
>>> from scipy.stats.contingency import expected_freq
>>> expected_freq(count)
array([[0.92307692, 1.07692308],
       [2.30769231, 2.69230769],
       [1.84615385, 2.15384615],
       [0.92307692, 1.07692308]])
>>> from scipy.stats.contingency import chi2_contingency
>>> chi2_contingency(count)
(2.135714285714286, 0.5447213779284998, 3, array([[0.92307692, 1.07692308],
       [2.30769231, 2.69230769],
       [1.84615385, 2.15384615],
       [0.92307692, 1.07692308]]))
>>> 
```
So probably makes the most sense over there rather than stand alone and perhaps some tests to illustrate/confirm behavior with round trip (e.g. get counts from `crosstab` and feed them into `chi2_contingency`, might make sense as another test. This demonstrates usage for the user who is willing to read the tests and provides another test of how the function is likely to be used in a workflow.

rlucas7 commented:
test failure in travis is unrelated:
https://travis-ci.org/github/scipy/scipy/jobs/636922075#L1278

matthew-brett commented on scipy/stats/_crosstab.py:
```diff
@@ -0,0 +1,154 @@
+
+def crosstab(*args, levels=None):
+    """
+    Create a table of the counts of the tuples in ``zip(*args)``.
```
Is there a more friendly way of phrasing this?   I expect most people looking for this function will be thinking of a 2 way contigency table.  How about

```
Return table giving counts for each possible unique combination of input ``*args``
```

?  As a taster?

WarrenWeckesser replied:
I rephrased the first line similar to what you suggested.  I tweaked it a bit to keep it under 80 characters.

matthew-brett commented on scipy/stats/_crosstab.py:
```diff
@@ -0,0 +1,154 @@
+    args : sequences
+        A sequence of sequences whose unique aligned elements are to be
+        counted.  The sequences in args must all be the same length.
+    levels : sequence, optional
```
I think it's worth pointing out that the default values for `levels` for input `i` are the output of `np.unique(args[i])`.   Specifically, that the order in which the values occur in `args` is irrelevant - the order in `levels` is sorted - for example, for strings, alphabetically.

WarrenWeckesser replied:
I added more about this to the docstring.

matthew-brett commented on scipy/stats/_crosstab.py:
```diff
@@ -0,0 +1,154 @@
+
+    Returns
+    -------
+    elements : tuple of numpy.ndarrays.
```
Am I right in thinking these will be the same as the passed or calculated values of `levels`?

WarrenWeckesser replied:
Yes.  I'll make a note of that in the docstring.

matthew-brett commented on scipy/stats/_crosstab.py:
```diff
@@ -0,0 +1,154 @@
+    >>> a = ['A', 'B', 'A', 'A', 'B', 'B', 'A', 'A', 'B', 'B']
+    >>> x = ['X', 'X', 'X', 'Y', 'Z', 'Z', 'Y', 'Y', 'Z', 'Z']
+    >>> (avals, xvals), count = crosstab(a, x)
+    >>> avals.tolist()
```
Why `.tolist()`?

WarrenWeckesser replied:
At the time, I thought the output looked nicer, but if a typical reader of the docstring has the same reaction as you, I shouldn't use it.  I removed it in the update.

matthew-brett commented on scipy/stats/_crosstab.py:
```diff
@@ -0,0 +1,154 @@
+    array([[1, 1, 0, 1],
+           [1, 4, 0, 1],
+           [0, 3, 0, 3]])
+    """
```
?

```
L0 = len(args[0])
if L0 == 0:
```

etc?

WarrenWeckesser replied:
@matthew-brett, I don't understand this comment/question.  Could you elaborate?

matthew-brett replied:
I just meant that, rather than recalculating `len(args[0])` multiple times, it will be very slightly quicker, and perhaps neater, to calculate once and put into a variable.  Very minor point.  It's just that whenever I see recalculation, my Spidey-sense tingles - so I have to suppress it with the argument to myself that it's so quick it probably doesn't matter.

WarrenWeckesser replied:
Gotcha, thanks.  I use both `len(args)` and `len(args[0])` in multiple places, so I'll define variables for those.

WarrenWeckesser replied:
Fixed in the update.

matthew-brett commented on scipy/stats/tests/test_crosstab.py:
```diff
@@ -0,0 +1,75 @@
+    (avals, bvals), count = crosstab(a, b)
+    assert_array_equal(avals, expected_avals)
+    assert_array_equal(bvals, expected_bvals)
+    assert_array_equal(count, expected_count)
```
Worth adding test where input arguments do not occur in `np.unique` order?

WarrenWeckesser replied:
I just tweaked the order of the first two elements in the second argument of the first test.  Does that cover this case sufficiently?

matthew-brett replied:
Yes, that's fine.

matthew-brett reviewed (commented):
Thanks much for this - and sorry to be slow to review.   I'd love to see this get in soon.  Please let me know what else I can do.

WarrenWeckesser commented:
I pushed an update that I think addresses all the review comments.

In the Notes in the docstring, I marked the function as new in 1.7.  I'd be happy to make this a last minute addition to 1.6, but I'm not expecting that.

I haven't figured how to resolve the problem reported in the `refguide_asv_check` test:
```
=======================
scipy.stats.contingency
=======================

Non-deprecated objects in __all__: 4
Objects in refguide: 0

ERROR: objects in scipy.stats.contingency.__all__ but not in refguide::

    crosstab

This issue can be fixed by adding these objects to
the function listing in __init__.py for this module

```
The module docstring in `scipy/stats/__init__.py` includes `contingency.crosstab`, along with `contingency.expected_freq` and `contingency.margins`.  The latter two have been listed like that for a long time, so I'm not sure what is causing the problem with `contingency.crosstab`.

As can be seen in the circleci artifact link, the docs build and `contingency.crosstab` is listed in the `stats` docs, and links correctly to its own page.

matthew-brett commented:
Thanks - that resolves all my questions.  I would love to see this get in for 1.6 - and it seems like a simple addition that should not cause any unexpected consequences.

WarrenWeckesser commented:
I fixed the refguide check problem, so all tests now pass.

matthew-brett commented:
@rlucas7 - would you mind casting an eye over this version?   Do you see any more problems to address?

mdhaber commented:
When this goes in, I'll use it in gh-12531.

rlucas7 commented:
I can take a look sometime later in the week or this weekend. 

Sincerely, 

-Lucas Roberts 

> On Dec 4, 2020, at 4:02 AM, Matthew Brett <notifications@github.com> wrote:
> 
> ﻿
> @rlucas7 - would you mind casting an eye over this version? Do you see any more problems to address?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

matthew-brett commented:
@rlucas7 - that would be great - thanks.

WarrenWeckesser commented:
> When this goes in, I'll use it in [gh-12531](https://github.com/scipy/scipy/pull/12531).

I think you meant gh-12653.

mdhaber commented:
Yes, I did.

rlucas7 reviewed (approved):
@WarrenWeckesser there are a couple edge cases that could have tests added on a follow up. 

The code coverage tool is pointing out the `ValueError` and `TypeError` branches. 
I'll open an issue w/a good first issue tag for those, that's a good one for someone to get their feet wet w/unit testing.

rlucas7 commented:
Squashing+merging.

@WarrenWeckesser  thanks for the PR & @matthew-brett thanks for the review

rlucas7 commented:
I left a 1.7.0 tag on this one, I think @tylerjereddy has already branched 1.6.0 release and there doesn't appear to be anything time sensitive in this one to get into 1.6.0 but let me know if I'm mistaken.

WarrenWeckesser commented:
Thanks @rlucas7 and all reviewers!  You're not mistaken, this will be in 1.7.

WarrenWeckesser commented:
I started new [release notes page for 1.7](https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.7.0) in the wiki, and added a note about `crosstab` to it.
